### PR TITLE
Pipes Rebecca/Lauren SCRABBLE 

### PR DESCRIPTION
**Pull request by laurenelee**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  It makes sense because we don't want an instance of the score. Every time we call on  the score method on the class we don't want it to be specific to any instance.  |
|   Describe an example of a test you wrote for a _nominal case_. | We wrote many for expected scores, calculating manually exactly what we expected a word equal.   	|
|   Describe an example of a test you wrote for an _edge case_. |  We wrote tests to raise ArgumentError if the player enters an empty string for their name. 	|
|   What was your final test coverage percentage?	|  100%   	|
|   Is there a specific piece of code you'd like feedback on?	|   We would like to learn more about the best way to tackle the optionals. We did the first one by changing the get_word method in the Wave-3-game.rb and changing the play(word) method in the player.rb file. 	|

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
+require 'pry'
+
+module Scrabble
+  class Scoring
+
+    LETTER_VALUES = {
+      'a' => 1,
+      'b' => 3,
+      'c' => 3,
```
Comment (droberts-sea) [inline lib/scoring.rb:9]: Good choice of data structure to hold this information

File: lib/scoring.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        score += LETTER_VALUES[char.downcase]
+      end
+      score += 50 if word.length == 7
+      return score
+    end
+
+    def self.tie(word, max_word)
+      if ((word.length == 7) && (max_word.length != 7)) || ((word.length < max_word.length) && (max_word.length != 7))
+        max_word = word
+        return max_word.downcase
```
Comment (droberts-sea) [inline lib/scoring.rb:51]: Good call breaking this out into its own method.

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
+require_relative 'spec_helper'
+
+describe 'Scrabble' do
+  describe 'Scoring' do
+
+    before do
+      @game1 = Scrabble::Scoring.new
+    end
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:8]: Why are you initializing this here? Isn't one of the big benefits of using all those `self` methods that you don't need an instance of `Scrabble::Scoring`?

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    end # new instance of scoring
+
+    describe "self.score " do
+      it "CAT returns 5" do
+        Scrabble::Scoring.score("cAt").must_equal 5
+      end # CAT logic
+
+      it "7 letter words receive 50 point bonus" do 
+        Scrabble::Scoring.score("laurenl").must_equal 57
+      end #50 point bonus
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:21]: What about an empty word, or an 8-letter word? You should test for all those `raise`es you wrote.

File: specs/scoring_spec.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+      it "tie will make shorter word win" do
+        array_of_words = ["be", "it", "cat", "date", "sit"]
+        Scrabble::Scoring.highest_score_from(array_of_words).must_equal "cat"
+      end #short word tie win
+
+      it "tie 7 letter word win" do
+        array_of_words = ["laurend", "dog", "cat", "zzzzzx", "sit"]
+        Scrabble::Scoring.highest_score_from(array_of_words).must_equal "laurend"
+      end # 7 letter tie
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:39]: Good coverage of ties here. Other cases I'd be interested in seeing:

- An empty array
- An array with only one word

File: specs/player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+describe 'Scrabble' do
+  describe 'Player' do
+
+    before do
+      @player = Scrabble::Player.new("Ada")
+    end
+    #mocking @tilebag to simulate test because we need to know exactly what is in the tile bag!
+    class TestTileBag
+      def initialize
+        @default_tiles = {
```
Comment (droberts-sea) [inline specs/player_spec.rb:12]: Good use of mocking!

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            proc{ @player.play("two words") }.must_raise ArgumentError
+          end
+
+          it "words with more than 7 letters can't be player" do
+            proc {@player.play("zazazaza")}.must_raise ArgumentError
+          end
+
+          it "empty strings cannot be played" do
+            proc {@player.play("")}.must_raise ArgumentError
+          end
```
Comment (droberts-sea) [inline specs/player_spec.rb:98]: Good coverage of failure cases here.

File: specs/player_spec.rb [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+            @player.draw_tiles(TestTileBag.new)
+
+            @player.play("zzzzzzz")
+            @player.play("z")
+            @player.play("zz").must_equal false
+          end
+
+          it "returns highest scoring word" do
+            @player.draw_tiles(TestTileBag.new)
+            @player.play("zzzzz")
```
Comment (droberts-sea) [inline specs/player_spec.rb:110]: Should this be inside a new `describe` block?

File: specs/player_spec.rb [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+                end
+
+                it "doesn't win if less than 100" do
+                  @player.draw_tiles(TestTileBag.new)
+                  @player.play("zzzz")
+                  @player.send(:won?).must_equal false
+                end
+              end
+            end
+          end
```
Comment (droberts-sea) [inline specs/player_spec.rb:171]: I think the nesting on your `describe` blocks might be a little off here.

File: lib/tilebag.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        'z' => 1
+      }
+      @tiles_remaining = 98
+    end
+
+
+    def draw_tiles(num)
+      @default_tiles.delete_if {|k, v| v == 0}
+      tiles_drawn = @default_tiles.keys.sample(num)
+      tiles_drawn.each do |drawn_tile|
```
Comment (droberts-sea) [inline lib/tilebag.rb:43]: This will only let you draw one of each letter!

File: lib/tilebag.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        's' => 4,
+        't' => 6,
+        'u' => 4,
+        'v' => 2,
+        'w' => 2,
+        'x' => 1,
+        'y' => 2,
+        'z' => 1
+      }
+      @tiles_remaining = 98
```
Comment (droberts-sea) [inline lib/tilebag.rb:36]: Rather than having tile remaining be an instance variable, you should just write a method to count the number of tiles in the bag. That way there's no way for them to get out of sync.

File: lib/tilebag.rb [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  class TileBag
+
+    attr_reader :default_tiles, :tiles_remaining, :tile_bag
+
+
+    def initialize
+      @default_tiles = {
+        'a' => 9,
+        'b' => 2,
+        'c' => 2,
```
Comment (droberts-sea) [inline lib/tilebag.rb:11]: This is a good, readable data structure for storing the starting tiles. However, it means you either need a really complicated loop below in `draw_tiles`, or you run into the bug that your code exhibits. It also makes it complicated to figure out how many tiles are left.

It might be cleaner to walk through this hash here and add the right amount of each letter to a big array, then draw from that instead.

File: lib/tilebag.rb (hunk 10 above)
Comment (droberts-sea) [reply lib/tilebag.rb:43]: A test that would catch this: try to draw 27 tiles.

File: specs/tilebag_spec.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+      it "returns num random tiles" do
+        Scrabble::TileBag.new.draw_tiles(1).length.must_equal 1
+      end
+
+      it "must return an array of tiles" do
+        Scrabble::TileBag.new.draw_tiles(1).must_be_kind_of String
+      end
+
+
+      describe "tiles_remaining" do
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:29]: You should close the previous `describe` block before starting a new one.

File: specs/tilebag_spec.rb [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    end
+
+    describe "draw_tiles" do
+      it "returns num random tiles" do
+        Scrabble::TileBag.new.draw_tiles(1).length.must_equal 1
+      end
+
+      it "must return an array of tiles" do
+        Scrabble::TileBag.new.draw_tiles(1).must_be_kind_of String
+      end
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:26]: This test doesn't match its name! The name claims it's looking for an array, but you check for a string.

Comment (droberts-sea) [reply specs/tilebag_spec.rb:26]: You're missing a couple of interesting edge cases here:
- What happens if you attempt to draw 0 tiles?
- Are you able to draw down to 0 tiles remaining?
- What happens when you attempt to draw more tiles than are in the bag?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  yes  |
|   Both Teammates contributed to the codebase	|  yes  |
|   Regular Commits with meaningful commit messages  |  yes  |
|  **Scoring class**	|   	|
|  `score` method	|  yes  |
|   Uses appropriate data structure to store the letter score	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  see inline comment  |
|  `highest_scoring_word` method	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `highest_scoring_word` edge cases (ties, empty array)	|  see inline comment  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  yes  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  yes  |
|  Returns `true` or `false` from `won?` method |  yes  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  yes  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  see inline comment  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  yes, but there's a bug  |
| `tiles_remaining` method returns a count	|  yes  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|  see inline comment  |

Good work - this code is well-organized and easy to read!

Good work overall, especially on product code. There are a few places where you've got holes in your test coverage, which I've tried to outline below. The pattern I'm seeing is that you should make sure to explore both sides of any edge cases. For example, in `TileBag#draw_tiles`, it's important to consider what happens if you attempt to draw more tiles than are in the bag (this should fail), and to check that you can draw the last tile in the bag (this should work). You can look at your code and say "of course that would work", but until you test it you can't know for sure.

Becoming a master tester is not easy, and this is the first project where you've had to build tests around complex logic, so don't worry if it's not yet where we want it to be. We'll have plenty of time to practice going forward. Do make sure to spend time thinking about it for Hotel, particularly around how ranges of dates interact, and let one of us know if you want us to check your work.
